rpcs3/rpcs3qt/update_manager.cpp

```diff
@@ -0,0 +1,513 @@
+	std::string tmp_folder = fs::get_config_dir() + m_tmp_folder;
+	fs::create_dir(tmp_folder);
+
+	for (UInt32 i = 0; i < db.NumFiles; i++)
+	{
+		size_t offset           = 0;
+		size_t outSizeProcessed = 0;
+		size_t len;
+		unsigned isDir = SzArEx_IsDir(&db, i);
+		len            = SzArEx_GetFileNameUtf16(&db, i, NULL);
+
+		if (len > PATH_MAX)
+		{
+			LOG_ERROR(GENERAL, "[Auto-updater] 7z decoder error: filename longer than PATH_MAX");
+			error_free7z();
+			return false;
+		}
+
+		SzArEx_GetFileNameUtf16(&db, i, temp_u16);
+		// Simplistic conversion to UTF-8
+		for (size_t index = 0; index < len; index++)
+		{
+			if (temp_u16[index] > 0xFF)
+			{
+				LOG_ERROR(GENERAL, "[Auto-updater] 7z decoder error: Failed to convert utf-16 to utf-8");
+				error_free7z();
+				return false;
+			}
+
+			temp_u8[index] = (u8)temp_u16[index];
+		}
+		std::string name((char*)temp_u8);
+
+		// Assumes rpcs3.exe is in the config directory(TODO: should be executable directory maybe?)
+		name = fs::get_config_dir() + name;
+
+		if (!isDir)
+		{
+			res = SzArEx_Extract(&db, &lookStream.vt, i, &blockIndex, &outBuffer, &outBufferSize, &offset, &outSizeProcessed, &allocImp, &allocTempImp);
+			if (res != SZ_OK)
+				break;
+		}
+
+		if (size_t pos = name.find_last_of('/'); pos != std::string::npos)
+		{
+			LOG_TRACE(GENERAL, "[Auto-updater] Creating path: %s", name.substr(0, pos));
+			fs::create_path(name.substr(0, pos));
+		}
+
+		if (isDir)
+		{
+			LOG_TRACE(GENERAL, "[Auto-updater] Creating dir: %s", name);
+			fs::create_dir(name);
+			continue;
+		}
+
+		fs::file outfile(name, fs::read + fs::write + fs::create + fs::trunc);
+		if (!outfile)
+		{
+			// File failed to open, probably because in use, rename existing file and try again
+			const auto pos = name.find_last_of('/');
+			std::string filename;
+			if (pos == std::string::npos)
+				filename = name;
+			else
+				filename = name.substr(pos + 1);
+
+			// Moving to temp is not an option on windows as it will fail if the disk is different
+			// So we create a folder in config dir and move stuff there
+			const std::string rename_target = tmp_folder + filename;
+			LOG_TRACE(GENERAL, "[Auto-updater] Renaming %s to %s", name, rename_target);
+			if (!fs::rename(name, rename_target, true))
+			{
+				LOG_ERROR(GENERAL, "[Auto-updater] Failed to rename %s to %s", name, rename_target);
+				res = SZ_ERROR_FAIL;
+				break;
+			}
+
+			outfile.open(name, fs::read + fs::write + fs::create + fs::trunc);
+			if (!outfile)
+			{
+				LOG_ERROR(GENERAL, "[Auto-updater] can not open output file %s", name);
+				res = SZ_ERROR_FAIL;
+				break;
+			}
+		}
+
+		if (outfile.write(outBuffer + offset, outSizeProcessed) != outSizeProcessed)
+		{
+			LOG_ERROR(GENERAL, "[Auto-updater] can not write output file: %s", name);
+			res = SZ_ERROR_FAIL;
+			break;
+		}
+		outfile.close();
+	}
+
+	error_free7z();
+	if (res)
+		return false;
+
+	const std::string path_to_exe = fs::get_config_dir() + "rpcs3.exe";
```

Comment:
User can modify config dir with the `RPCS3_CONFIG_DIR` enviroment variable, it is not guaranteed to be the same as rpcs3 executable location.

Reply:
Changed to using Emulator::GetEmuDir().

Reply:
hmm nvm I didn't see it was a private member of Emulator, well GetModuleFileName it is then.

Reply:
Or I guess I can just make it public, I don't see a reason not to.